crates/pera-core/src/dwallet_mpc/mod.rs

```diff
@@ -0,0 +1,409 @@
+        session_id: deserialized_event.session_id.bytes,
+        initiating_user_address: deserialized_event.initiator,
+        mpc_round: MPCRound::BatchedPresign(deserialized_event.batch_size),
+    }
+}
+
+pub(crate) fn advance<P: AsynchronouslyAdvanceable>(
+    session_id: CommitmentSizedNumber,
+    party_id: PartyID,
+    access_threshold: &WeightedThresholdAccessStructure,
+    messages: Vec<HashMap<PartyID, MPCMessage>>,
+    public_input: P::PublicInput,
+    private_input: P::PrivateInput,
+) -> DwalletMPCResult<mpc::AsynchronousRoundResult<Vec<u8>, Vec<u8>, Vec<u8>>> {
+    let messages = deserialize_mpc_messages(messages)?;
+
+    let res = P::advance(
+        session_id,
+        party_id,
+        access_threshold,
+        messages,
+        Some(private_input),
+        &public_input,
+        &mut rand_core::OsRng,
+    )
+    .map_err(|e| DwalletMPCError::TwoPCMPCError(format!("{:?}", e)))?;
+
+    Ok(match res {
+        mpc::AsynchronousRoundResult::Advance {
+            malicious_parties,
+            message,
+        } => mpc::AsynchronousRoundResult::Advance {
+            malicious_parties,
+            message: bcs::to_bytes(&message)?,
+        },
+        mpc::AsynchronousRoundResult::Finalize {
+            malicious_parties,
+            private_output: _,
+            public_output,
+        } => {
+            let public_output: P::PublicOutputValue = public_output.into();
+            let public_output = bcs::to_bytes(&public_output)?;
+            mpc::AsynchronousRoundResult::Finalize {
+                malicious_parties,
+                private_output: Vec::new(),
+                public_output,
+            }
+        }
+    })
+}
+
+/// Deserializes the messages received from other parties for the next advancement.
+/// Any value that fails to deserialize is considered to be sent by a malicious party.
+/// Returns the deserialized messages or an error including the IDs of the malicious parties.
+fn deserialize_mpc_messages<M: DeserializeOwned + Clone>(
+    messages: Vec<HashMap<PartyID, MPCMessage>>,
+) -> DwalletMPCResult<Vec<HashMap<PartyID, M>>> {
+    let mut deserialized_results = Vec::new();
+    let mut malicious_parties = Vec::new();
+
+    for message_batch in &messages {
+        let mut valid_messages = HashMap::new();
+
+        for (party_id, message) in message_batch {
+            match bcs::from_bytes::<M>(&message) {
+                Ok(value) => {
+                    valid_messages.insert(*party_id, value);
+                }
+                Err(_) => {
+                    malicious_parties.push(*party_id);
+                }
+            }
+        }
+
+        if !valid_messages.is_empty() {
+            deserialized_results.push(valid_messages);
+        }
+    }
+
+    if !malicious_parties.is_empty() {
+        Err(DwalletMPCError::MaliciousParties(malicious_parties))
+    } else {
+        Ok(deserialized_results)
+    }
```

Comment:
I have a few problems with this function.

Firstly, it uses mutable data structures for no good reason. As a guideline, avoid mutable data structures and opt for functional programming whenever it is possible and makes sense.

Second, you return an error in the case of malicious parties identified; then what? You DOS the system as the operation failed and we cannot `advance`. We can never let one (or `< t`) parties DOS the system. The only way you can avoid this with this design, is by calling this function, receiving an error, parsing the malicious parties out of the error, filter this function's input, and call it again - ridiculously complicated.

Third, I already implemented a helper function just for this: `.handle_invalid_messages_async()`. This function does two things for you: first, it allows you to simply operate over an iterator of `(PartyID, Result<>)`, which is convenient as you don't have to worry about unwrapping the results and handling errors and converting between data structures (list and hash maps). Second, it returns a tuple, the first is all the malicious parties (those with `Err`) and the second is the filtered and unwrapped mapping (in your case: the deserialized messages which are valid.)

Returning a tuple like this guarantees success, and that no small number of parties would be able to DOS. Then, the caller can handle the malicious parties as they wish, but still keep running (e.g. `advance()`ing.)